File: algorithms/algorithms-cplusplus/leetcode/Question_0802_Find_Eventual_Safe_States.cc
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
    int n = graph.size();
    // 0 未遍历 1 正在遍历 2 已遍历结束
    vector<int> color(n, 0), res;
    for (int i = 0; i < n; i++) {
      if (dfs(graph, i, color)) {
        res.push_back(i);
      }
    }
     return res;
  }

 private:
  bool dfs(vector<vector<int>>& graph, int i, vector<int>& color) {
    if (color[i] > 0) {
      return color[i] == 2;
    }
    color[i] = 1;
    for (int j : graph[i]) {
      if (color[j] == 1 || !dfs(graph, j, color)) {
        return false;
      }
    }
    color[i] = 2;
    return true;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0802_Find_Eventual_Safe_States.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Question_0802_Find_Eventual_Safe_States {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int n = graph.length;
        // 0 未遍历 1 正在遍历 2 已遍历结束
        int[] color = new int[n];
        List<Integer> res = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (dfs(i, color, graph)) {
                res.add(i);
            }
        }
        return res;
    }

    public boolean dfs(int node, int[] color, int[][] graph) {
        if (color[node] > 0) {
            return color[node] == 2;
        }
        color[node] = 1;
        for (int nei : graph[node]) {
            if (color[nei] == 1 || !dfs(nei, color, graph)) {
                return false;
            }
        }
        color[node] = 2;
        return true;
    }
}
